File: solutions/javascript/freelancer-rates/1/freelancer-rates.js
// @ts-check

/**
 * The day rate, given a rate per hour
 *
 * @param {number} ratePerHour
 * @returns {number} the rate per day
 */

const billableDaysPerMonth = 22;
const hoursPerDay = 8;

export function dayRate(ratePerHour) {
	return hoursPerDay * ratePerHour;
}

/**
 * Calculates the number of days in a budget, rounded down
 *
 * @param {number} budget: the total budget
 * @param {number} ratePerHour: the rate per hour
 * @returns {number} the number of days
 */
export function daysInBudget(budget, ratePerHour) {
	return Math.floor(budget /  dayRate(ratePerHour));
}

/**
 * Calculates the discounted rate for large projects, rounded up
 *
 * @param {number} ratePerHour
 * @param {number} numDays: number of days the project spans
 * @param {number} discount: for example 20% written as 0.2
 * @returns {number} the rounded up discounted rate
 */
export function priceWithMonthlyDiscount(ratePerHour, numDays, discount) {
	const nonDiscountedDays = numDays % billableDaysPerMonth;
	const discountedDays = numDays - nonDiscountedDays;
	const discountedDaysRate = discountedDays * (1 - discount);
	const baseDailyPrice = dayRate(ratePerHour);
	return Math.ceil(baseDailyPrice * (nonDiscountedDays + discountedDaysRate));
}

File: solutions/elixir/freelancer-rates/1/lib/freelancer_rates.ex
defmodule FreelancerRates do
  @working_hours_per_day 8.0
  @working_days_per_month 22.0

  def daily_rate(hourly_rate) do
    @working_hours_per_day * hourly_rate
  end

  def apply_discount(before_discount, discount) do
    before_discount * (1 - discount / 100)
  end

  def monthly_rate(hourly_rate, discount) do
    base_rate = daily_rate(hourly_rate) * @working_days_per_month
    ceil(apply_discount(base_rate, discount))
  end

  def days_in_budget(budget, hourly_rate, discount) do
    Float.floor(budget / apply_discount(daily_rate(hourly_rate), discount), 1)
  end
end
